### PR TITLE
Derive quantizers using linear models

### PR DESCRIPTION
**Pull request by YaLTeR**
This is the result of derf's and my investigation into quantizer averaging ([av1-metric.pdf](http://dgql.org/~derf/av1-metric.pdf), Section 2.2). Six quantizers are derived from `log_target_q`, the average quantizer, using linear models built on test encodes. The notebook with plots and weight derivation can be found [here](https://github.com/YaLTeR/rav1e/blob/8e4095dbe40a66c41a140577ba3c6d6f7b1427d8/tools/analyse-vstats.ipynb).

AWCY for [4:2:0](https://beta.arewecompressedyet.com/?job=master-a811e9ef43e36ca9334146ca3a993a964ae45168&job=q-from-q-bar%402020-08-11T07%3A40%3A26.668Z), [4:2:2 (based on an earlier commit)](https://beta.arewecompressedyet.com/?job=5cdb1c50-4%3A2%3A2%402020-08-05T17%3A31%3A18.940Z&job=q-from-q-bar-dist-scale-2-4%3A2%3A2%402020-08-10T06%3A43%3A31.065Z) and [4:4:4 (based on an earlier commit)](https://beta.arewecompressedyet.com/?job=5cdb1c50-4%3A4%3A4%402020-08-06T05%3A33%3A28.076Z&job=q-from-q-bar-dist-scale-2-4%3A4%3A4%402020-08-10T06%3A43%3A19.868Z).

Additional latest latest master 4:2:0 results from _the secret AWCY instance_:
|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.6889 | -1.6590 | -1.0713 |  -0.3524 | -1.1467 | -0.9218 |    -1.2484 |

One of the questions is what to do with the data collection and analysis branch, currently at https://github.com/YaLTeR/rav1e/tree/luma_chroma_collect1_rebased. I feel the data collection changes there are too annoying to integrate with a feature flag. Derf has suggested to just push it as a branch on the main repo.

### CODE REVIEW

**Comment by shssoichiro** [comment]
There seems to be a pretty considerable hit to encoding speed from this. Is this reproducible locally, or just an artifact of AWCY being wonky?

**Comment by YaLTeR** [comment]
Updated as per the comment. New results:
- 4:2:0
    |    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
    |    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
    | -0.7100 | -1.7853 | -1.4440 |  -0.3603 | -1.1473 | -0.9403 |    -1.2956 |
- 4:2:2
    |   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
    |   ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
    | 0.0695 | -1.2680 | -1.3371 |   0.1597 | -0.2710 | -0.1917 |    -0.4929 |
- 4:4:4
    |   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
    |   ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
    | 0.0853 | -0.4476 |  0.4962 |   0.1721 | -0.0029 |  0.0419 |    -0.1086 |

File: src/rate.rs [hunk 1]
```diff
@@ -77,6 +77,30 @@ const DQP_Q57: &[i64; FRAME_NSUBTYPES] = &[
...
+];
+#[rustfmt::skip]
+const Q_MODEL_MUL: [i64; 3] = [
+  // 4:2:0
+  0x8A0_50DD,
+  // 4:2:2
+  0x887_7666,
+  // 4:4:4
+  0x8D4_A712,
+];
```
Comment (barrbrain) [inline src/rate.rs:102]: We may add entries for `ChromaSampling::Cs400` and remove its exclusion below.
```suggestion
const Q_MODEL_ADD: [i64; 4] = [
  // 4:2:0
  -0x24_4FE7_ECB3_DD90,
  // 4:2:2
  -0x37_41DA_38AD_0924,
  // 4:4:4
  -0x70_83BD_A626_311C,
  // 4:0:0
  0,
];
#[rustfmt::skip]
const Q_MODEL_MUL: [i64; 4] = [
  // 4:2:0
  0x8A0_50DD,
  // 4:2:2
  0x887_7666,
  // 4:4:4
  0x8D4_A712,
  // 4:0:0
  0,
];
```

**Review by barrbrain — APPROVED** [review]
This looks ready to merge, given that we should confirm results for non-8-bit and intra.

**Comment by YaLTeR** [comment]
I've pushed the data collection branch into https://github.com/xiph/rav1e/tree/quantizer-weight-data-collection with `analyse-vstats.ipynb` [moved over](https://github.com/xiph/rav1e/blob/d02bdbd3b0b7b2cb9fc301031cc6a4e67a567a5c/doc/quantizer-weight-analysis.ipynb) to the doc folder. I've also added it into this PR in the same location so it exists on master too. I've additionally updated the link in the comment to point to the new location.

File: src/rate.rs [hunk 2]
```diff
@@ -703,22 +731,33 @@ fn chroma_offset(
...
-      * ((log_target_q_u as f64) * Q57_SQUARE_EXP_SCALE).exp();
-    let lambda_v = (::std::f64::consts::LN_2 / 6.0)
-      * ((log_target_q_v as f64) * Q57_SQUARE_EXP_SCALE).exp();
+
+    let log_q = [log_q_y, log_q_u, log_q_v];
+    let mut dist_scale = [0.; 3];
+    for plane in 0..3 {
+      dist_scale[plane] =
+        bexp64((log_target_q - log_q[plane]) * 2 + q57(16)) as f64 / 65536.;
+    }
```
Comment (barrbrain) [inline src/rate.rs:760]: For the record, we could apply the map-and-collect pattern like so:
```suggestion
    let dist_scale = [log_q_y, log_q_u, log_q_v]
      .iter()
      .map(|q| bexp64((log_target_q - q) * 2 + q57(16)) as f64 / 65536.)
      .collect::<ArrayVec<_>>()
      .into_inner()
      .unwrap();
```

Comment (YaLTeR) [reply src/rate.rs:760]: How about:
```rust
    let scale = |q| bexp64((log_target_q - q) * 2 + q57(16)) as f64 / 65536.;
    let dist_scale = [scale(log_q_y), scale(log_q_u), scale(log_q_v)];
```

**Pull request merged**